File: warehouse/erp_the20/repo/notification.repository.js
// repo/notification.repository.js

const { Op } = require("sequelize");

module.exports = function createNotificationRepo({ Notification }) {
  async function createNotification({
    title,
    recipients = null,
    to_user = null,
    payload = null,
    object_type = "",
    object_id = "",
    channel = Notification.Channel?.INAPP ?? 0,
    delivered = true,
  }) {
    const now = new Date();
    return Notification.create({
      title,
      recipients: recipients ? [...recipients] : null,
      to_user,
      payload: payload || {},
      object_type,
      object_id,
      channel,
      delivered,
      delivered_at: delivered ? now : null,
      attempt_count: delivered ? 1 : 0,
    });
  }

  async function listByUser(user_id, limit = 200) {
    return Notification.findAll({
      where: {
        [Op.or]: [
          { to_user: user_id },
          // recipients: ARRAY<Integer> on Postgres
          { recipients: { [Op.contains]: [user_id] } },
        ],
      },
      order: [["created_at", "DESC"]],
      limit,
    });
  }

  return { createNotification, listByUser };
};

File: warehouse/erp_the20/repositories/notification_repository.py
from typing import Iterable, Optional, Dict, Any, List
from django.utils import timezone
from django.db.models import Q, QuerySet
from erp_the20.models import Notification

def create_notification(
    title: str,
    *,
    recipients: Optional[Iterable[int]] = None,
    to_user: Optional[int] = None,
    payload: Optional[Dict[str, Any]] = None,
    object_type: str = "",
    object_id: str = "",
    channel: int = Notification.Channel.INAPP,
    delivered: bool = True,
) -> Notification:
    return Notification.objects.create(
        title=title,
        recipients=list(recipients) if recipients else None,
        to_user=to_user,
        payload=payload or {},
        object_type=object_type,
        object_id=object_id,
        channel=channel,
        delivered=delivered,
        delivered_at=timezone.now() if delivered else None,
        attempt_count=1 if delivered else 0,
    )

def list_by_user(user_id: int, limit: int = 200) -> QuerySet:
    return (
        Notification.objects
        .filter(Q(to_user=user_id) | Q(recipients__contains=[user_id]))
        .order_by("-created_at")[:limit]
    )
